pkg/controller/networkpolicy/validate.go

```diff
@@ -119,95 +182,72 @@ func (v *NetworkPolicyValidator) Validate(ar *admv1.AdmissionReview) *admv1.Admi
 }
 
 // validateAntreaPolicy validates the admission of a Antrea NetworkPolicy CRDs
-func (v *NetworkPolicyValidator) validateAntreaPolicy(op admv1.Operation, tier string, ingress, egress []secv1alpha1.Rule) (string, bool) {
+func (v *NetworkPolicyValidator) validateAntreaPolicy(curObj, oldObj interface{}, op admv1.Operation, userInfo authenticationv1.UserInfo) (string, bool) {
 	allowed := true
 	reason := ""
 	switch op {
-	case admv1.Create, admv1.Update:
-		if isUnique := v.validateRuleName(ingress, egress); !isUnique {
-			allowed = false
-			reason = "rules names are not unique, or policy has duplicate rules, or collision occurred in generated rule names"
-			break
-		}
-		// "tier" must exist before referencing.
-		if tier == "" || staticTierSet.Has(tier) {
-			// Empty Tier name corresponds to default Tier
-			break
+	case admv1.Create:
+		for _, val := range v.antreaPolicyValidators {
+			reason, allowed = val.createValidate(curObj, userInfo)
+			if !allowed {
+				return reason, allowed
+			}
 		}
-		if ok := v.tierExists(tier); !ok {
-			allowed = false
-			reason = fmt.Sprintf("tier %s does not exist", tier)
+	case admv1.Update:
+		for _, val := range v.antreaPolicyValidators {
+			reason, allowed = val.updateValidate(curObj, oldObj, userInfo)
+			if !allowed {
+				return reason, allowed
+			}
 		}
 	case admv1.Delete:
-		// Delete of Antrea Policies have no validation
-		allowed = true
-	}
-	return reason, allowed
-}
-
-// validateRuleName validates if the name of each rule is unique within a policy
-func (v *NetworkPolicyValidator) validateRuleName(ingress, egress []secv1alpha1.Rule) bool {
-	uniqueRuleName := sets.NewString()
-	isUnique := func(rules []secv1alpha1.Rule) bool {
-		for _, rule := range rules {
-			if uniqueRuleName.Has(rule.Name) {
-				return false
+		// Delete of Antrea Policies have no validation. This will be an
+		// empty for loop.
+		for _, val := range v.antreaPolicyValidators {
+			reason, allowed = val.deleteValidate(curObj, userInfo)
```

Comment:
Should it be oldObj too, although it has no validation now?
I'm not sure which object has the deleted object in delete case, curObj or oldObj. But `validateTier` is using oldObj.

Reply:
this should be oldObj too..

Reply:
since there is no validation on delete.. there is no functional effect and hence no e2e test case today.. but oldObj should be passed here in case there are validations in future